ddsrouter_core/src/cpp/participant/implementations/rtps/CommonParticipant.cpp

```diff
@@ -83,50 +89,92 @@ void CommonParticipant::onParticipantDiscovery(
 }
 
 template<class DiscoveryInfoKind>
-types::Endpoint CommonParticipant::create_endpoint_from_info_(
+types::Endpoint CommonParticipant::create_common_endpoint_from_info_(
         DiscoveryInfoKind& info)
 {
     // Parse GUID
     types::Guid info_guid;
     info_guid = info.info.guid();
 
-    // Parse QoS
-    types::DurabilityKind info_durability_kind = info.info.m_qos.m_durability.durabilityKind();
-    types::ReliabilityKind info_reliability_kind;
+    // Parse TopicQoS
+    types::TopicQoS discovered_topic_qos;
+    // Durability
+    discovered_topic_qos.durability_qos = info.info.m_qos.m_durability.durabilityKind();
+    // Reliability
     if (info.info.m_qos.m_reliability.kind == fastdds::dds::BEST_EFFORT_RELIABILITY_QOS)
     {
-        info_reliability_kind = fastrtps::rtps::BEST_EFFORT;
+        discovered_topic_qos.reliability_qos = fastrtps::rtps::BEST_EFFORT;
     }
     else if (info.info.m_qos.m_reliability.kind == fastdds::dds::RELIABLE_RELIABILITY_QOS)
     {
-        info_reliability_kind = fastrtps::rtps::RELIABLE;
+        discovered_topic_qos.reliability_qos = fastrtps::rtps::RELIABLE;
     }
     else
     {
         utils::tsnh(
             utils::Formatter() <<
                 "Invalid ReliabilityQoS value found while parsing DiscoveryInfo for Endpoint creation.");
     }
-    types::QoS info_qos(info_durability_kind, info_reliability_kind);
+    // Set Topic with Partitions
+    discovered_topic_qos.use_partitions = !info.info.m_qos.m_partition.empty();
+    // Set Topic with ownership
+    discovered_topic_qos.ownership_qos = info.info.m_qos.m_ownership.kind;
+
+    // Parse specific QoS of the entity
+    types::SpecificWriterQoS specific_qos;
+    if (discovered_topic_qos.has_partitions())
+    {
+        specific_qos.partitions = info.info.m_qos.m_partition;
+    }
+    // NOTE: ownership is only for Writer
 
     // Parse Topic
-    types::RealTopic info_topic(std::string(info.info.topicName()), std::string(info.info.typeName()),
-            info.info.topicKind() == eprosima::fastrtps::rtps::TopicKind_t::WITH_KEY);
+    types::DdsTopic info_topic(std::string(info.info.topicName()), std::string(info.info.typeName()));
+    info_topic.keyed = info.info.topicKind() == eprosima::fastrtps::rtps::TopicKind_t::WITH_KEY;
+    // Set qos as set, but fuzzy
+    info_topic.topic_qos = discovered_topic_qos;
+    info_topic.topic_qos.set_level(utils::FuzzyLevelValues::fuzzy_level_fuzzy);
+
+    return types::Endpoint(
+        types::EndpointKind::invalid,
+        info_guid,
+        info_topic,
+        this->id_nts_());
+}
 
-    // Create Endpoint
-    if (std::is_same<DiscoveryInfoKind, fastrtps::rtps::ReaderDiscoveryInfo>::value)
```

Comment:
They are not better, they are just mandatory.
These two classes have not a related hierarchy (I do not know why) and they have different values at some point (in this case `m_ownershipStrength`. Thus, this could not be done in a common parametrized method as this value would be tried to be compiled for both classes, and one of them has not this variable.
Thus, the possibilities were:
1. to create two methods that do basically the same but for 2 different classes
2. maintain the common and separate the specific part in different methods, so there are 3 methods.